### DISCUSSION TITLE
Relational field update fails when user's role has granular permission

### DISCUSSION THREAD

**Discussion by darioguarascio** [directus/directus#10771] [Archive]
A user with granular permission fails to add new rows in a table that has a M2O relationship. 
As admin role, with all permissions set, everything works well.

The custom role has all permission set 

### To Reproduce

In table `A` there is a `website` field Many to One to `website` table and a simple text field `name`

I am accessing directus as user that has read/write/delete/update permission on table A where `website -> name Equals x`

When I try to add a new row to table `A`, selecting an entry from the `website` table with the `M2O` selector, I get the following error

```
{"errors":[{"message":"\"website\" must be of type object","extensions":{"code":"FAILED_VALIDATION","field":"website"}}]}
```

the data being posted is the following:
```
{
    "name": "test",
    "website": 37
}
```

### Errors Shown

![image](https://user-images.githubusercontent.com/3661363/147655912-c44490d6-6fc8-4882-8e48-f4e6da86dc21.png)

**Comment by darioguarascio** [comment]
I have to add: updating an already existing row, with the restricted role, works ok (at least for non-relational fields)

**Comment by rijkvanzanten** [comment]
Heya!

I believe this is caused by some confusion in how the validation rules operate. When you configure the validation to be

```json
{
  "website": {
    "id": {
      "_eq": 37
    }
  }
}
```

the API requires an object to be sent in that exact shape, eg

```json
{
  "website": {
    "id": 37
  }
}
```

However, that in turn also means that the following is considered invalid:

```json
{
  "website": 37
}
```

even though the end result is similar.

To get around this, you can configure your permissions to be an "or" using:

```json
{
  "_or": [
    {
      "website": {
        "_eq": 37
      }
    },
    {
      "website": {
        "id": {
          "_eq": 37
        }
      }
    }
  ]
}
```

---

This configuration issue arises pretty often actually, so more than happy to hear any and all thoughts on how we could improve this setup 🙂

**Reply by darioguarascio** [reply]
@rijkvanzanten thank you for the hint.

Trying it out, I get your point. Unfortunately, by my test it happens that only the `{"_or":[{"website":{"_eq":37}}]}` works.
the data being sent in the update request refers to ids only.


there are 2 options I think:
- send a whole object (IE: web.name = x ) and match it against the validation rules
- send the id only (as for now), load the validation rule, expand a given id into an object that can be compared to the validation rule, the compare it.

**Reply by rijkvanzanten** [reply]
> the data being sent in the update request refers to ids only.

That depends on if an item was "selected" or "created", and if the "selected" item was updated as well 🤔 

> send a whole object (IE: web.name = x ) and match it against the validation rules

Yeah this makes sense to me! It seems like we could avoid a lot of confusion here by just having the app send the id as the object structure at all times. The only thing we'd have to figure out is that we make sure we don't trigger an update on that item, based on the existence of the payload. We might have to add a read to ensure there's no diff before processing the item (cc @Oreilles)

**Reply by gdcl** [reply]
thanks for filing this discussion item - ran into the same exact issue trying to implement some rbac control on logged in user test for an item with an M2O relation to a directus user. Workaround works for me. 

Behavior certainly is confusing and not straightforward to troubleshoot without google (and @darioguarascio's discussion item). 

I like the way the Directus App suggests to create the filter in the UI - it is pretty straightforward - so the fix towards ensuring all CRUD's send the payload consistently so they match a singular filter expression seems the more intuitive approach.

**Reply by highvibesonly** [reply]
Thanks for the discussion about validation. I'd like to follow up this one here, since I don't know a better suitable location since I guess it is not a software issue.

I have a table `localgroups` (primary id is Integer), that has members and admins. The members can add and delete themselves to/from the group. The admins (managers) can add/delete others as well.

This is realized with 2x M2M relationships to the `directus_users`:
- from `localgroup.administrators` to `directus_users.localgroup_administrator` using the junction table `localgroups_directus_users_administrators`.
- from `localgroups.members` to `directus_users.localgroup_members` using the junction table `localgroups_directus_users_members`.

__Permissions__

- Create-permission of table `localgroup` is set to `id` and `members`.
- Create-permission of junction-table `localgroup_directus_users_members` is all checked (id, localgroup_id, directus_user_id).

__Validation__
- When i set the create-permission for new entries to `localgroups_directus_users_members` (adding a user to a group) as followed:

```
{"_and":[{"localgroup_id":{"_in":["$CURRENT_USER.localgroup_administrator.localgroup_id"]}}]}
```

__Error__
On making the PATCH request with the body:

```
{
  "create": [
    {
      "localgroup_id": "6",
      "directus_users_id": {
        "id": "48cac1d5-3d13-4a5a-8774-7b44dc1068ff"
      }
    }
  ],
  "update": [],
  "delete": []
}
```

... I get the error:

```
{
  "message": "Validation failed for field \"localgroup_id\". Value has to be \"6\".",
  "extensions": {
    "code": "FAILED_VALIDATION",
    "field": "localgroup_id",
    "type": "eq",
    "valid": 6
  }
}
```

I tried to send the same request with the id as number, but this didnt work. What am I missing here?

Thanks in advance!
